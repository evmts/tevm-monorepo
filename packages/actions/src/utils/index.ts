export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
