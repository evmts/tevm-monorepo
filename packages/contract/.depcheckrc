ignores: [
  "@evmts/core",
  "@evmts/ts-plugin",
  "zod"
]
skip-missing: true
