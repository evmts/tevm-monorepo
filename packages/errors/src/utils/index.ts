export type { DecodeFunctionDataError } from './DecodeFunctionDataError.js'
export type { EncodeFunctionReturnDataError } from './EncodeFunctionReturnDataError.js'
