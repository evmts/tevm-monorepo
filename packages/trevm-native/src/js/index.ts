import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};