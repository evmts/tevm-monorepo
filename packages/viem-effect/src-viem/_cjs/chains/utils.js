'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.formattersOptimism =
	exports.parseTransactionCelo =
	exports.serializersCelo =
	exports.serializeTransactionCelo =
	exports.formattersCelo =
		void 0
const formatters_js_1 = require('./celo/formatters.js')
Object.defineProperty(exports, 'formattersCelo', {
	enumerable: true,
	get: function () {
		return formatters_js_1.formattersCelo
	},
})
const serializers_js_1 = require('./celo/serializers.js')
Object.defineProperty(exports, 'serializeTransactionCelo', {
	enumerable: true,
	get: function () {
		return serializers_js_1.serializeTransactionCelo
	},
})
Object.defineProperty(exports, 'serializersCelo', {
	enumerable: true,
	get: function () {
		return serializers_js_1.serializersCelo
	},
})
const parsers_js_1 = require('./celo/parsers.js')
Object.defineProperty(exports, 'parseTransactionCelo', {
	enumerable: true,
	get: function () {
		return parsers_js_1.parseTransactionCelo
	},
})
const formatters_js_2 = require('./optimism/formatters.js')
Object.defineProperty(exports, 'formattersOptimism', {
	enumerable: true,
	get: function () {
		return formatters_js_2.formattersOptimism
	},
})
//# sourceMappingURL=utils.js.map
