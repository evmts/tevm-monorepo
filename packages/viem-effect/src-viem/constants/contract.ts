export const aggregate3Signature = '0x82ad56cb'
