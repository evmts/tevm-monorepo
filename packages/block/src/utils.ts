import { EthjsAddress, keccak256, toType as originalToType, toBytes } from '@tevm/utils'

/**
 * Creates a zero-filled buffer of the specified length
 * @param {number} length - The length of the buffer
 * @returns {Uint8Array} A zero-filled Uint8Array
 */
export function zeros(length: number): Uint8Array {
	return new Uint8Array(length)
}

/**
 * Creates the zero address
 * @returns {EthjsAddress} The zero address
 */
export function createZeroAddress() {
	return new EthjsAddress(new Uint8Array(20))
}

/**
 * Creates an address from a public key
 * @param {Uint8Array} publicKey - The public key
 * @returns {EthjsAddress} The address
 */
export function createAddressFromPublicKey(publicKey: Uint8Array): EthjsAddress {
	// Create address from public key by taking the last 20 bytes of keccak256 hash
	const addressBytes = keccak256(publicKey, 'bytes').slice(-20)
	return new EthjsAddress(addressBytes)
}

/**
 * Wrapper for toType that handles string inputs properly
 * @param {any} input - The input value
 * @param {number} outputType - The output type
 * @returns {any} The converted value
 */
export function safeToType(input: any, outputType: number): any {
	if (input === null || input === undefined) {
		return undefined
	}

	// Handle string inputs by converting to bytes first
	if (typeof input === 'string' && outputType === 2) {
		const bytes = toBytes(input)
		return originalToType(bytes, outputType as any)
	}

	return originalToType(input, outputType as any)
}

/**
 * Gets the v value from a signature
 * @param {any} signature - The signature object
 * @returns {bigint} The v value
 */
export function getSignatureV(signature: any): bigint {
	// Handle both old format (v) and new format (recovery)
	if (signature.v !== undefined) {
		return BigInt(signature.v)
	}
	if (signature.recovery !== undefined) {
		return BigInt(signature.recovery) + 27n
	}
	throw new Error('Invalid signature format: missing v or recovery')
}
