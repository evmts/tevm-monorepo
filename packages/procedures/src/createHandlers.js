import { chainIdHandler } from '@tevm/actions'
import { MethodNotSupportedError } from '@tevm/errors'
import { anvilDropTransactionJsonRpcProcedure } from './anvil/anvilDropTransactionProcedure.js'
import { anvilDumpStateJsonRpcProcedure } from './anvil/anvilDumpStateProcedure.js'
import { anvilGetAutomineJsonRpcProcedure } from './anvil/anvilGetAutomineProcedure.js'
import { anvilImpersonateAccountJsonRpcProcedure } from './anvil/anvilImpersonateAccountProcedure.js'
import { anvilLoadStateJsonRpcProcedure } from './anvil/anvilLoadStateProcedure.js'
import { anvilResetJsonRpcProcedure } from './anvil/anvilResetProcedure.js'
import { anvilSetBalanceJsonRpcProcedure } from './anvil/anvilSetBalanceProcedure.js'
import { anvilSetCodeJsonRpcProcedure } from './anvil/anvilSetCodeProcedure.js'
import { anvilSetCoinbaseJsonRpcProcedure } from './anvil/anvilSetCoinbaseProcedure.js'
import { anvilSetNonceJsonRpcProcedure } from './anvil/anvilSetNonceProcedure.js'
import { anvilSetStorageAtJsonRpcProcedure } from './anvil/anvilSetStorageAtProcedure.js'
import { anvilStopImpersonatingAccountJsonRpcProcedure } from './anvil/anvilStopImpersonatingAccountProcedure.js'
import { callProcedure } from './call/callProcedure.js'
import { debugTraceCallJsonRpcProcedure } from './debug/debugTraceCallProcedure.js'
import { debugTraceTransactionJsonRpcProcedure } from './debug/debugTraceTransactionProcedure.js'
import { dumpStateProcedure } from './dumpstate/dumpStateProcedure.js'
import { blockNumberProcedure } from './eth/blockNumberProcedure.js'
import { chainIdProcedure } from './eth/chainIdProcedure.js'
import { ethBlobBaseFeeJsonRpcProcedure } from './eth/ethBlobBaseFeeProcedure.js'
import { ethCallProcedure } from './eth/ethCallProcedure.js'
import { ethCoinbaseJsonRpcProcedure } from './eth/ethCoinbaseProcedure.js'
import { ethEstimateGasJsonRpcProcedure } from './eth/ethEstimateGasProcedure.js'
import { ethGetBlockByHashJsonRpcProcedure } from './eth/ethGetBlockByHashProcedure.js'
import { ethGetBlockByNumberJsonRpcProcedure } from './eth/ethGetBlockByNumberProcedure.js'
import { ethGetBlockTransactionCountByHashJsonRpcProcedure } from './eth/ethGetBlockTransactionCountByHashProcedure.js'
import { ethGetBlockTransactionCountByNumberJsonRpcProcedure } from './eth/ethGetBlockTransactionCountByNumberProcedure.js'
import { ethGetFilterChangesProcedure } from './eth/ethGetFilterChangesProcedure.js'
import { ethGetFilterLogsProcedure } from './eth/ethGetFilterLogsProcedure.js'
import { ethGetLogsProcedure } from './eth/ethGetLogsProcedure.js'
import { ethGetTransactionByBlockHashAndIndexJsonRpcProcedure } from './eth/ethGetTransactionByBlockHashAndIndexProcedure.js'
import { ethGetTransactionByBlockNumberAndIndexJsonRpcProcedure } from './eth/ethGetTransactionByBlockNumberAndIndexProcedure.js'
import { ethGetTransactionByHashJsonRpcProcedure } from './eth/ethGetTransactionByHashProcedure.js'
import { ethGetTransactionCountProcedure } from './eth/ethGetTransactionCountProcedure.js'
import { ethGetTransactionReceiptJsonRpcProcedure } from './eth/ethGetTransactionReceiptProcedure.js'
import { ethNewBlockFilterProcedure } from './eth/ethNewBlockFilterProcedure.js'
import { ethNewFilterJsonRpcProcedure } from './eth/ethNewFilterProcedure.js'
import { ethNewPendingTransactionFilterProcedure } from './eth/ethNewPendingTransactionFilterProcedure.js'
import { ethProtocolVersionJsonRpcProcedure } from './eth/ethProtocolVersionProcedure.js'
import { ethSendRawTransactionJsonRpcProcedure } from './eth/ethSendRawTransactionProcedure.js'
import { ethSendTransactionJsonRpcProcedure } from './eth/ethSendTransactionProcedure.js'
import { ethUninstallFilterJsonRpcProcedure } from './eth/ethUninstallFilterProcedure.js'
import { gasPriceProcedure } from './eth/gasPriceProcedure.js'
import { getBalanceProcedure } from './eth/getBalanceProcedure.js'
import { getCodeProcedure } from './eth/getCodeProcedure.js'
import { getStorageAtProcedure } from './eth/getStorageAtProcedure.js'
import { getAccountProcedure } from './getaccount/getAccountProcedure.js'
import { loadStateProcedure } from './loadstate/loadStateProcedure.js'
import { mineProcedure } from './mine/mineProcedure.js'
import { scriptProcedure } from './script/scriptProcedure.js'
import { setAccountProcedure } from './setaccount/setAccountProcedure.js'

/**
 * @typedef {ReturnType<typeof createHandlers>} RequestHandlers
 */

/**
 * @internal
 * Creates a mapping of methods to jsonrpc request handlers
 * @param {import('@tevm/node').TevmNode} client
 * @returns Request handlers
 * @example
 */
export const createHandlers = (client) => {
	const tevmHandlers = {
		tevm_call: callProcedure(client),
		/**
		 * @param {any} request
		 */
		tevm_contract: (request) => {
			const err = new MethodNotSupportedError(
				'UnsupportedMethodError: tevm_contract is not supported. Encode the contract arguments and use tevm_call instead.',
			)
			return /**@type any*/ ({
				id: /** @type any*/ (request).id,
				method: request.method,
				jsonrpc: '2.0',
				error: {
					code: err._tag,
					message: err.message,
				},
			})
		},
		tevm_getAccount: getAccountProcedure(client),
		tevm_setAccount: setAccountProcedure(client),
		tevm_script: scriptProcedure(client),
		tevm_dumpState: dumpStateProcedure(client),
		tevm_loadState: loadStateProcedure(client),
		tevm_miner: mineProcedure(client),
	}

	const ethHandlers = {
		eth_blockNumber: blockNumberProcedure(client),
		eth_chainId: chainIdProcedure(client),
		eth_call: ethCallProcedure(client),
		eth_getCode: getCodeProcedure(client),
		eth_getStorageAt: getStorageAtProcedure(client),
		eth_gasPrice: gasPriceProcedure(client),
		eth_getBalance: getBalanceProcedure(client),
		eth_coinbase: ethCoinbaseJsonRpcProcedure(client),
		eth_mining:
			/**
			 * @param {any} request}
			 */
			(request) => {
				return {
					result: client.status === 'MINING',
					method: request.method,
					jsonrpc: '2.0',
					...(request.id ? { id: request.id } : {}),
				}
			},
		eth_syncing:
			/**
			 * @param {any} request}
			 */
			(request) => {
				return {
					result: client.status === 'SYNCING',
					method: request.method,
					jsonrpc: '2.0',
					...(request.id ? { id: request.id } : {}),
				}
			},
		eth_sendTransaction: ethSendTransactionJsonRpcProcedure(client),
		eth_sendRawTransaction: ethSendRawTransactionJsonRpcProcedure(client),
		eth_estimateGas: ethEstimateGasJsonRpcProcedure(client),
		eth_getTransactionReceipt: ethGetTransactionReceiptJsonRpcProcedure(client),
		eth_getLogs: ethGetLogsProcedure(client),
		eth_getBlockByHash: ethGetBlockByHashJsonRpcProcedure(client),
		eth_getBlockByNumber: ethGetBlockByNumberJsonRpcProcedure(client),
		eth_getBlockTransactionCountByHash: ethGetBlockTransactionCountByHashJsonRpcProcedure(client),
		eth_getBlockTransactionCountByNumber: ethGetBlockTransactionCountByNumberJsonRpcProcedure(client),
		eth_getTransactionByHash: ethGetTransactionByHashJsonRpcProcedure(client),
		eth_getTransactionByBlockHashAndIndex: ethGetTransactionByBlockHashAndIndexJsonRpcProcedure(client),
		eth_getTransactionByBlockNumberAndIndex: ethGetTransactionByBlockNumberAndIndexJsonRpcProcedure(client),
		eth_protocolVersion: ethProtocolVersionJsonRpcProcedure(),
		eth_getTransactionCount: ethGetTransactionCountProcedure(client),
		eth_newFilter: ethNewFilterJsonRpcProcedure(client),
		eth_getFilterLogs: ethGetFilterLogsProcedure(client),
		eth_newBlockFilter: ethNewBlockFilterProcedure(client),
		eth_uninstallFilter: ethUninstallFilterJsonRpcProcedure(client),
		eth_getFilterChanges: ethGetFilterChangesProcedure(client),
		eth_newPendingTransactionFilter: ethNewPendingTransactionFilterProcedure(client),
		eth_blobBaseFee: ethBlobBaseFeeJsonRpcProcedure(client),
	}

	const anvilHandlers = {
		anvil_setCode: anvilSetCodeJsonRpcProcedure(client),
		anvil_setBalance: anvilSetBalanceJsonRpcProcedure(client),
		anvil_setNonce: anvilSetNonceJsonRpcProcedure(client),
		anvil_setChainId: chainIdHandler(client),
		anvil_getAutomine: anvilGetAutomineJsonRpcProcedure(client),
		anvil_setCoinbase: anvilSetCoinbaseJsonRpcProcedure(client),
		anvil_mine: mineProcedure(client),
		anvil_reset: anvilResetJsonRpcProcedure(client),
		anvil_dropTransaction: anvilDropTransactionJsonRpcProcedure(client),
		anvil_dumpState: anvilDumpStateJsonRpcProcedure(client),
		anvil_loadState: anvilLoadStateJsonRpcProcedure(client),
		anvil_setStorageAt: anvilSetStorageAtJsonRpcProcedure(client),
		anvil_impersonateAccount: anvilImpersonateAccountJsonRpcProcedure(client),
		anvil_stopImpersonatingAccount: anvilStopImpersonatingAccountJsonRpcProcedure(client),
	}
	const tevmAnvilHandlers = Object.fromEntries(
		Object.entries(anvilHandlers).map(([key, value]) => {
			return [key.replace('anvil', 'tevm'), value]
		}),
	)
	const ganacheHandlers = Object.fromEntries(
		Object.entries(anvilHandlers).map(([key, value]) => {
			return [key.replace('anvil', 'ganache'), value]
		}),
	)
	const hardhatHandlers = Object.fromEntries(
		Object.entries(anvilHandlers).map(([key, value]) => {
			return [key.replace('anvil', 'hardhat'), value]
		}),
	)

	const debugHandlers = {
		debug_traceTransaction: debugTraceTransactionJsonRpcProcedure(client),
		debug_traceCall: debugTraceCallJsonRpcProcedure(client),
	}

	const allHandlers = {
		...tevmHandlers,
		...ethHandlers,
		...anvilHandlers,
		...tevmAnvilHandlers,
		...ganacheHandlers,
		...hardhatHandlers,
		...debugHandlers,
	}

	return allHandlers
}
