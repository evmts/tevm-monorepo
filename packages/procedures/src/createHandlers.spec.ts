import { ERC20 } from '@tevm/contract'
import { createTevmNode } from '@tevm/node'
import { numberToHex } from '@tevm/utils'
import { beforeEach, describe, expect, it } from 'vitest'
import { type RequestHandlers, createHandlers } from './createHandlers.js'

const ERC20_ADDRESS = `0x${'69'.repeat(20)}` as const

describe('createHandlers', () => {
	let client: any
	let handlers: RequestHandlers

	beforeEach(() => {
		client = createTevmNode()
		handlers = createHandlers(client)
	})

	it('should handle eth_chainId', async () => {
		const res = await handlers.eth_chainId({
			jsonrpc: '2.0',
			method: 'eth_chainId',
			id: 1,
		})
		expect(res).toMatchSnapshot()
	})

	it('should handle eth_call', async () => {
		const to = `0x${'69'.repeat(20)}` as const
		const res = await handlers.eth_call({
			jsonrpc: '2.0',
			method: 'eth_call',
			id: 1,
			params: [
				{
					from: to,
					to,
					data: '0x',
				},
				'latest',
			],
		})
		expect(res).toMatchSnapshot()
	})

	it('should handle eth_getBalance', async () => {
		const address = `0x${'69'.repeat(20)}` as const
		const res = await handlers.eth_getBalance({
			jsonrpc: '2.0',
			method: 'eth_getBalance',
			id: 1,
			params: [address, 'latest'],
		})
		expect(res).toMatchSnapshot()
	})

	it('should handle tevm_getAccount', async () => {
		await handlers.tevm_setAccount({
			jsonrpc: '2.0',
			method: 'tevm_setAccount',
			id: 1,
			params: [
				{
					address: ERC20_ADDRESS,
					deployedBytecode: ERC20.deployedBytecode,
					balance: numberToHex(420n),
					nonce: numberToHex(69n),
				},
			],
		})
		const res = await handlers.tevm_getAccount({
			jsonrpc: '2.0',
			method: 'tevm_getAccount',
			id: 1,
			params: [
				{
					address: ERC20_ADDRESS,
				},
			],
		})
		expect(res.error).toBeUndefined()
		expect(res).toMatchSnapshot
	})

	it('should handle eth_estimateGas', async () => {
		const to = `0x${'69'.repeat(20)}` as const
		const res = await handlers.eth_estimateGas({
			jsonrpc: '2.0',
			method: 'eth_estimateGas',
			id: 1,
			params: [
				{
					from: to,
					to,
					data: '0x',
				},
			],
		})
		expect(res).toMatchSnapshot()
	})

	it('should handle tevm_mine', async () => {
		const res = await handlers.tevm_miner({
			jsonrpc: '2.0',
			method: 'tevm_mine',
			id: 1,
			params: ['0x1', '0x1'],
		})
		expect(res.id).toBe(1)
		expect(res.error).toBeUndefined()
		expect(res.method).toBe('tevm_mine')
		expect(res.result?.blockHashes).toHaveLength(1)
	})

	it('should handle tevm_contract', async () => {
		const res = await handlers.tevm_contract({
			jsonrpc: '2.0',
			method: 'tevm_contract',
			id: 1,
		})
		expect(res).toMatchSnapshot()
	})

	// Add more tests for other handlers as needed...

	it('should handle eth_mining', async () => {
		const res = handlers.eth_mining({
			jsonrpc: '2.0',
			method: 'eth_mining',
			id: 1,
		})
		expect(res).toMatchSnapshot()
	})

	it('should handle eth_syncing', async () => {
		const res = handlers.eth_syncing({
			jsonrpc: '2.0',
			method: 'eth_syncing',
			id: 1,
		})
		expect(res).toMatchSnapshot()
	})
})
