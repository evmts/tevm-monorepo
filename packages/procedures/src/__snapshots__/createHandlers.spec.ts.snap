// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createHandlers should handle tevm_mine 1`] = `
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "tevm_mine",
  "result": {
    "blockHashes": [
      "0x3201277b6823bbcd45e1a1c4a334c31b083676c70cebb1f386139f6091ae8e49",
    ],
  },
}
`;

exports[`createHandlers should handle eth_chainId 1`] = `
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "eth_chainId",
  "result": "0x384",
}
`;

exports[`createHandlers should handle eth_call 1`] = `
{
  "error": {
    "code": -32000,
    "data": {
      "errors": [
        
"sender doesn't have enough funds to send tx. The upfront cost is: 210000000 and the sender's account (0x6969696969696969696969696969696969696969) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0x96171020921a6736a4e62d65c40b749609224fe43187d437cb37d19b3c011005 nonce=0 value=0 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
      ],
    },
    "message": 
"sender doesn't have enough funds to send tx. The upfront cost is: 210000000 and the sender's account (0x6969696969696969696969696969696969696969) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0x96171020921a6736a4e62d65c40b749609224fe43187d437cb37d19b3c011005 nonce=0 value=0 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
  },
  "id": 1,
  "jsonrpc": "2.0",
  "method": "eth_call",
}
`;

exports[`createHandlers should handle eth_getBalance 1`] = `
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "eth_getBalance",
  "result": "0x0",
}
`;

exports[`createHandlers should handle eth_estimateGas 1`] = `
{
  "error": {
    "code": -32000,
    "data": {
      "errors": [
        
"sender doesn't have enough funds to send tx. The upfront cost is: 210000000 and the sender's account (0x6969696969696969696969696969696969696969) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0x96171020921a6736a4e62d65c40b749609224fe43187d437cb37d19b3c011005 nonce=0 value=0 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
      ],
    },
    "message": 
"sender doesn't have enough funds to send tx. The upfront cost is: 210000000 and the sender's account (0x6969696969696969696969696969696969696969) only has: 0 -> block number=0 hash=0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454 hf=cancun baseFeePerGas=7 txs=0 uncles=0 -> tx type=2 hash=0x96171020921a6736a4e62d65c40b749609224fe43187d437cb37d19b3c011005 nonce=0 value=0 signed=true hf=cancun maxFeePerGas=7 maxPriorityFeePerGas=0)

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73

Docs: https://tevm.sh/reference/tevm/errors/classes/insufficientfundserror/
Details: /reference/tevm/errors/classes/insufficientfundserror/
Version: 1.1.0.next-73"
,
  },
  "id": 1,
  "jsonrpc": "2.0",
  "method": "eth_estimateGas",
}
`;

exports[`createHandlers should handle tevm_contract 1`] = `
{
  "error": {
    "code": "MethodNotSupported",
    "message": 
"UnsupportedMethodError: tevm_contract is not supported. Encode the contract arguments and use tevm_call instead.

Docs: https://tevm.sh/reference/tevm/errors/classes/methodnotsupportederror/
Version: 1.1.0.next-73"
,
  },
  "id": 1,
  "jsonrpc": "2.0",
  "method": "tevm_contract",
}
`;

exports[`createHandlers should handle eth_mining 1`] = `
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "eth_mining",
  "result": false,
}
`;

exports[`createHandlers should handle eth_syncing 1`] = `
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "eth_syncing",
  "result": false,
}
`;
